### PR TITLE
Feature/dynamic min fungible fees

### LINKED ISSUE #361: Dynamic Min Amounts per transfer
Currently, every fungible native token has an optional min amount that when you portal tokens in both directions, it verifies whether the amount is more than the minimum. This amount is in the lowest denomination of the token and currently a constant, which is set upon validator initialisation in `bridge.yml` config (ref https://github.com/LimeChain/hashport-validator/blob/main/docs/configuration.md `min_amount`.
The task is to dynamically calculate the min amount for each token, based on the work done from all validators.

When a portal transfer is found, each validator does the same work - it submits Hedera transactions based on the type of transfer, whether they will be Scheduled Transactions or Consensus Submit Messages. This work cost is multiplied by the number of validators, which is then considered the fee percentage that each token has upon portal transfer. As HBAR goes up / down against USD, the changing exchange rate will mean that a given transaction will cost fewer or more HBAR over time – resulting in a stable USD price. [Here](https://limechain.notion.site/Portal-Fees-in-USD-c7346c5894bb4e2db2e0f27137fa876f) is an example of calculated min amounts for each token.

Currently, Hedera has an account `0.0.112` that stores the current rate of 1 HBAR in USD, which is updated by `0.0.57` periodically (approx every hour).
e.g. `currentRate : 0.2616, nextRate : 0.2623, midnight-currentRate : 0.2631 midnight-nextRate : 0.2619`.
[Example query](https://testnet.mirrornode.hedera.com/api/v1/transactions?account.id=0.0.57&transactiontype=fileupdate&timestamp=gt:1641880836). The rates are encoded in `memo_base64`.

Tasks: 
- [ ] Calculate the minimum amounts for each token that can be bridged to another network in USD
* [Example of fees](https://limechain.notion.site/Portal-Fees-in-USD-c7346c5894bb4e2db2e0f27137fa876f)
* [Costs in Hedera](https://docs.hedera.com/guides/mainnet/fees#transaction-and-query-fees)
- [ ] Switch the minimum amount for each token to be in USD in `bridge.yml` configuration. 
- [ ] Create a watcher that gets the current rate of HBAR in USD from account `0.0.112`
* Given that the file is updated once an hour, you can get the rate every 10 mins, to ensure you will not update it later than expected.
* [Example query](https://testnet.mirrornode.hedera.com/api/v1/transactions?account.id=0.0.57&transactiontype=fileupdate&timestamp=gt:1641880836)
- [ ] Create a service, that converts the minimum amount for the token from USD to the given token. (e.g. you have the cost in USD and the token is WETH, you need to convert the amount to WETH in the lowest denomination).
* Have at least two providers - main and fallback, in case the main fails to get the rate. Example: Coingecko.
- [ ] Expose an API call `/api/v1/min-amounts/` that returns the minimum amounts for each token in their currency
Example: 
```json
{
  "HBAR": "11600000000" // 116 HBAR In the lowest denomination (tinybar)
  "WETH": "8000000000000000" // 0.008 WETH in the lowest denomination (wei). You can return either symbol, or an address, which you find the most unique.
}
```
- [ ] Update configuration & integration documentation, which includes the new API call and the updated configuration properties.

Additional articles:
https://hedera.com/blog/hedera-fee-model-for-transactions-queries

### PR DESCRIPTION
**Pull request by nnedkov-limechain**
**Detailed description**:
Implementing Dynamic Fungible Min Amounts. And refactoring Assets to be a service.

**Which issue(s) this PR fixes**:
Fixes #361

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: app/domain/client/contract.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package client
```
Comment (svetlio8) [inline app/domain/client/contract.go:1]: Do we need such a file - looks like an empty one?

File: app/clients/coin-gecko/web-api/client.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package web_api
```
Comment (svetlio8) [inline app/clients/coin-gecko/web-api/client.go:1]: missing license

File: app/clients/coin-gecko/web-api/client.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
```
Comment (svetlio8) [inline app/clients/coin-gecko/web-api/client.go:2]: missing license

File: app/clients/coin-market-cap/web-api/client.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package web_api
```
Comment (svetlio8) [inline app/clients/coin-market-cap/web-api/client.go:1]: missing license

File: app/clients/coin-market-cap/web-api/client.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
```
Comment (svetlio8) [inline app/clients/coin-market-cap/web-api/client.go:2]: missing license

File: app/clients/hedera/web-api/client.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package web_api
```
Comment (svetlio8) [inline app/clients/hedera/web-api/client.go:1]: missing license

File: app/domain/client/coin-gecko-web-api.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package client
```
Comment (svetlio8) [inline app/domain/client/coin-gecko-web-api.go:1]: missing license

File: app/domain/client/coin-gecko-web-api.go [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
```
Comment (svetlio8) [inline app/domain/client/coin-gecko-web-api.go:2]: missing license

File: app/domain/client/coin-market-cap-web-api.go [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package client
```
Comment (svetlio8) [inline app/domain/client/coin-market-cap-web-api.go:1]: missing license

File: app/domain/client/hedera-web-api.go [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package client
```
Comment (svetlio8) [inline app/domain/client/hedera-web-api.go:1]: missing license

File: app/domain/service/assets.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package service
```
Comment (svetlio8) [inline app/domain/service/assets.go:1]: missing license

File: app/helper/clients/clients.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package clients
```
Comment (svetlio8) [inline app/helper/clients/clients.go:1]: missing license

File: app/helper/coin-gecko/web-api.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package coin_gecko
```
Comment (svetlio8) [inline app/helper/coin-gecko/web-api.go:1]: looks like the missing license is valid for all newly created files

File: app/helper/http/get.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+		req.Header.Set(key, value)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("Error while sending request to server. Error: [%v]", err)
+		return responseBodyAsBytes, err
+	}
+
+	responseBodyAsBytes, _ = ioutil.ReadAll(resp.Body)
```
Comment (vectorman1) [inline app/helper/http/get.go:26]: Unclosed stream;
Unchecked error;
io/ioutil is deprecated;
If possible, avoid using ReadAll;

Maybe add an `interface{}` argument to for a struct ref to which you can
`json.NewDecoder(resp.Body).Decode(struct)` the response. 

If helpers for deserialization is an established pattern you can pass the deserialization function as well.

**Review by vectorman1 — CHANGES_REQUESTED** [review]

File: config/bridge.go [hunk 15]
```diff
@@ -17,17 +17,20 @@
...
 )
 
 type Bridge struct {
-	TopicId string
-	Hedera  *BridgeHedera
-	EVMs    map[uint64]BridgeEvm
-	Assets  Assets
+	TopicId          string
+	Hedera           *BridgeHedera
+	Assets           service.Assets
```
Comment (failfmi) [inline config/bridge.go:30]: Since Assets is no longer a config, but a service, wouldn't be best to remove it from here?

File: config/bridge.go [hunk 16]
```diff
@@ -61,39 +73,50 @@ func NewBridge(bridge parser.Bridge) Bridge {
...
-			hederaFeePercentages, hederaNftFees := LoadHederaFees(value.Tokens)
+			hederaFeePercentages, hederaNftFees := LoadHederaFees(networkInfo.Tokens)
 			config.Hedera.FeePercentages = hederaFeePercentages
 			config.Hedera.NftFees = hederaNftFees
-			continue
-		}
-		config.EVMs[key] = BridgeEvm{
-			RouterContractAddress: value.RouterContractAddress,
-			Tokens:                make(map[string]Token),
+
```
Comment (failfmi) [inline config/bridge.go:100]: nit: empty line

File: config/bridge.go [hunk 17]
```diff
@@ -61,39 +73,50 @@ func NewBridge(bridge parser.Bridge) Bridge {
...
-			config.EVMs[key].Tokens[name] = Token{Networks: value.Networks}
+
+		for name, tokenInfo := range networkInfo.Tokens.Fungible {
+			config.CoinGeckoIds[networkId][name] = tokenInfo.CoinGeckoId
+			config.CoinMarketCapIds[networkId][name] = tokenInfo.CoinMarketCapId
+			if networkId == constants.HederaNetworkId {
+				config.Hedera.Tokens[name] = NewHederaTokenFromToken(tokenInfo)
+			}
 		}
+
```
Comment (failfmi) [inline config/bridge.go:119]: nit: empty line

File: node.yml [hunk 18]
```diff
@@ -16,9 +18,11 @@
...
+#      api_address: "https://api.coingecko.com/api/v3/"
+#    coin_market_cap:
+#      api_key:
+#      api_address: "https://pro-api.coinmarketcap.com/v2/cryptocurrency/"
 #  monitoring:
 #    enable: false
 #    dashboard_polling: 15 # in minutes
-#  log_level: info
-#  port: 5200
-#  validator: true
```
Comment (failfmi) [inline node.yml:24]: I think these should be visible here.

File: app/services/contracts/service.go [hunk 19]
```diff
@@ -203,3 +198,12 @@ func (bsc *Service) RemoveDecimals(amount *big.Int, asset string) (*big.Int, err
...
 }
+
+func (bsc *Service) TokenFeeData() (struct {
+	ServiceFeePercentage *big.Int
+	FeesAccrued          *big.Int
+	PreviousAccrued      *big.Int
+	Accumulator          *big.Int
+}, error) {
+	return bsc.contract.TokenFeeData(nil, common.Address{})
+}
```
Comment (failfmi) [inline app/services/contracts/service.go:209]: This is not used anywhere.

File: app/process/watcher/price/watcher.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	// there will be no handler, so the q is to implement the interface
+	go pw.beginWatching()
+
+}
+
+func (pw *Watcher) beginWatching() {
+	pw.pricingService.FetchAndUpdateUsdPrices(false)
+
+	time.Sleep(10 * time.Minute)
+}
```
Comment (failfmi) [inline app/process/watcher/price/watcher.go:50]: Doesn't this execute only once?

Comment (failfmi) [reply app/process/watcher/price/watcher.go:50]: Can we add some logs when this executes every 10 mins

File: app/process/watcher/price/watcher.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+	"time"
+)
+
+type Watcher struct {
+	pricingService service.Pricing
+	logger         *log.Entry
+}
+
+func NewWatcher(pricingService service.Pricing) *Watcher {
+
```
Comment (failfmi) [inline app/process/watcher/price/watcher.go:33]: empty line

File: app/process/watcher/price/watcher.go [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+	return &Watcher{
+		pricingService: pricingService,
+		logger:         config.GetLoggerFor("Price Watcher"),
+	}
+}
+
+func (pw *Watcher) Watch(q qi.Queue) {
+	// there will be no handler, so the q is to implement the interface
+	go pw.beginWatching()
+
```
Comment (failfmi) [inline app/process/watcher/price/watcher.go:43]: empty line

File: app/process/watcher/prometheus/watcher.go [hunk 23]
```diff
@@ -66,27 +68,29 @@ func NewWatcher(
...
-		},
-	})
-	operatorBalanceGauge = prometheusService.CreateGaugeIfNotExists(prometheus.GaugeOpts{
-		Name: constants.OperatorAccountAmountName,
-		Help: constants.OperatorAccountAmountHelp,
-		ConstLabels: prometheus.Labels{
-			constants.AccountMetricLabelKey: configuration.Node.Clients.Hedera.Operator.AccountId,
-		},
-	})
+	if prometheusService.GetIsMonitoringEnabled() {
```
Comment (failfmi) [inline app/process/watcher/prometheus/watcher.go:71]: Why is this needed? The watcher is not registered if this is not enabled.

File: app/process/watcher/prometheus/watcher.go [hunk 24]
```diff
@@ -99,14 +103,11 @@ func NewWatcher(
...
 		assetsMetrics:             assetsMetrics,
+		assetsService:             assetsService,
 	}
 }
 
 func (pw Watcher) Watch(q qi.Queue) {
-	if !pw.prometheusService.GetIsMonitoringEnabled() {
-		pw.logger.Warnf("Tried to executed Prometheus watcher, when monitoring is not enabled.")
-		return
-	}
```
Comment (failfmi) [inline app/process/watcher/prometheus/watcher.go:109]: ?

File: app/process/watcher/transfer/watcher.go [hunk 25]
```diff
@@ -279,8 +286,17 @@ func (ctw Watcher) createFungiblePayload(transactionID string, receiver string,
 			err)
 		return nil, err
 	}
-	if properAmount.Cmp(nativeAsset.MinAmount) < 0 {
-		return nil, errors.New(fmt.Sprintf("Transfer Amount [%s] is less than Minimum Amount [%s].", properAmount, nativeAsset.MinAmount))
+
+	tokenPriceInfo, exist := ctw.pricingService.GetTokenPriceInfo(asset.ChainId, nativeAsset.Asset)
+	if !exist {
+		errMsg := fmt.Sprintf("[%s] - Couldn't get price info in USD for asset [%s].", transactionID, nativeAsset.Asset)
+		ctw.logger.Errorf(errMsg)
```
Comment (failfmi) [inline app/process/watcher/transfer/watcher.go:293]: No need to log this here, it will log the error later.

File: app/process/watcher/transfer/watcher.go [hunk 26]
```diff
@@ -279,8 +286,17 @@ func (ctw Watcher) createFungiblePayload(transactionID string, receiver string,
...
+		return nil, errors.New(errMsg)
+	}
+
+	amountInUsd := decimalHelper.GetAmountInUsd(tokenPriceInfo.UsdPrice, properAmount, ctw.assetsService, nativeAsset)
+	if amountInUsd.Cmp(tokenPriceInfo.MinAmountInUsdWithFee) < 0 {
+		return nil, errors.New(fmt.Sprintf("[%s] - Transfer Amount [%s] is less than Minimum Amount [%s].", transactionID, amountInUsd, nativeAsset.MinFeeAmountInUsd))
```
Comment (failfmi) [inline app/process/watcher/transfer/watcher.go:299]: Wouldn't it be better to only compare amounts here rather than doing calculations every time?

File: app/process/watcher/evm/watcher.go [hunk 27]
```diff
@@ -486,8 +516,17 @@ func (ew *Watcher) handleLockLog(eventLog *router.RouterLock, q qi.Queue) {
...
+	nativeAsset := ew.assetsService.FungibleNativeAsset(sourceChainId, token)
+	tokenPriceInfo, exist := ew.pricingService.GetTokenPriceInfo(sourceChainId, nativeAsset.Asset)
+	if !exist {
+		ew.logger.Errorf("[%s] - Couldn't get price info in USD for asset [%s].", eventLog.Raw.TxHash, nativeAsset.Asset)
+		return
+	}
+
+	amountInUsd := decimalHelper.GetAmountInUsd(tokenPriceInfo.UsdPrice, properAmount, ew.assetsService, nativeAsset)
+	if amountInUsd.Cmp(tokenPriceInfo.MinAmountInUsdWithFee) < 0 {
+		ew.logger.Errorf("[%s] - Transfer Amount [%s] less than Minimum Amount [%s].", eventLog.Raw.TxHash, amountInUsd, nativeAsset.MinFeeAmountInUsd)
```
Comment (failfmi) [inline app/process/watcher/evm/watcher.go:529]: Wouldn't it be better to only compare amounts here rather than doing calculations every time?

File: app/process/watcher/evm/watcher.go [hunk 28]
```diff
@@ -467,16 +502,11 @@ func (ew *Watcher) handleLockLog(eventLog *router.RouterLock, q qi.Queue) {
...
+	wrappedAsset := ew.assetsService.NativeToWrapped(token, sourceChainId, targetChainId)
 	if wrappedAsset == "" {
 		ew.logger.Errorf("[%s] - Failed to retrieve native asset of [%s].", eventLog.Raw.TxHash, eventLog.Token)
 		return
 	}
-	nativeAsset := ew.mappings.FungibleNativeAsset(sourceChainId, token)
-	if eventLog.Amount.Cmp(nativeAsset.MinAmount) < 0 {
-		ew.logger.Errorf("[%s] - Transfer Amount [%s] less than Minimum Amount [%s].", eventLog.Raw.TxHash, eventLog.Amount, nativeAsset.MinAmount)
-		return
-	}
```
Comment (failfmi) [inline app/process/watcher/evm/watcher.go:479]: The check has to be done here otherwise using properAmount later might have a reduced value compared to the native token.

File: app/domain/client/coin-gecko.go [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package client
+
+import "github.com/shopspring/decimal"
+
+type CoinGecko interface {
+	// GetUsdPrices Returns USD prices for requested tokens
+	GetUsdPrices(idsByNetworkAndAddress map[uint64]map[string]string) (pricesByNetworkAndAddress map[uint64]map[string]decimal.Decimal, err error)
+}
```
[the remark this answers is not in the document]
Comment (nnedkov-limechain) [reply app/domain/client/coin-gecko.go:24]: Yep, I was thinking about doing it. Good Idea.

File: app/process/watcher/prometheus/watcher.go (hunk 24 above)
Comment (nnedkov-limechain) [reply app/process/watcher/prometheus/watcher.go:109]: This is by mistake will revert it.

File: app/process/watcher/prometheus/watcher.go (hunk 23 above)
Comment (nnedkov-limechain) [reply app/process/watcher/prometheus/watcher.go:71]: because if to do only 1 check instead of calling each of the methods and to return from inside. It is more efficient like that.

File: app/process/watcher/evm/watcher.go (hunk 27 above)
Comment (nnedkov-limechain) [reply app/process/watcher/evm/watcher.go:529]: Yep fixing it now.

File: app/process/watcher/evm/watcher.go [hunk 30]
```diff
@@ -486,8 +494,16 @@ func (ew *Watcher) handleLockLog(eventLog *router.RouterLock, q qi.Queue) {
...
-		ew.logger.Errorf("[%s] - Insufficient amount provided: Event Amount [%s] and Proper Amount [%s].", eventLog.Raw.TxHash, eventLog.Amount, properAmount)
+
+	nativeAsset := ew.assetsService.FungibleNativeAsset(sourceChainId, token)
+	tokenPriceInfo, exist := ew.pricingService.GetTokenPriceInfo(sourceChainId, nativeAsset.Asset)
+	if !exist {
+		ew.logger.Errorf("[%s] - Couldn't get price info in USD for asset [%s].", eventLog.Raw.TxHash, nativeAsset.Asset)
+		return
+	}
+
+	if properAmount.Cmp(tokenPriceInfo.MinAmountWithFee) < 0 {
```
Comment (failfmi) [inline app/process/watcher/evm/watcher.go:505]: `properAmount` -> `eventLog.Amount`

File: app/process/watcher/prometheus/watcher.go (hunk 23 above)
Comment (svetlio8) [reply app/process/watcher/prometheus/watcher.go:71]: @nnedkov-limechain I think @failfmi is referring to the check on init of the Prometheus watcher in [main.go](https://github.com/LimeChain/hashport-validator/blob/feature/dynamic-min-fungible-fees/cmd/main.go#L103)

File: app/process/watcher/prometheus/watcher.go [hunk 31]
```diff
@@ -241,6 +217,10 @@ func getMetricData(
 }
 
 func (pw Watcher) setMetrics() {
+	if !pw.prometheusService.GetIsMonitoringEnabled() {
```
Comment (svetlio8) [inline app/process/watcher/prometheus/watcher.go:220]: I think we have the same case here as on row [71](https://github.com/LimeChain/hashport-validator/blob/feature/dynamic-min-fungible-fees/app/process/watcher/prometheus/watcher.go#L71) -the if in [main.go](https://github.com/LimeChain/hashport-validator/blob/feature/dynamic-min-fungible-fees/cmd/main.go#L103)

File: app/process/watcher/prometheus/watcher.go [hunk 32]
```diff
@@ -300,6 +280,10 @@ func (pw Watcher) prepareAndSetAssetMetric(networkId uint64,
 	bridgeAccount *model.AccountsResponse,
 	isNative bool,
 ) {
+	if !pw.prometheusService.GetIsMonitoringEnabled() {
```
Comment (svetlio8) [inline app/process/watcher/prometheus/watcher.go:283]: There is the same check in the setMetrics function. Do we need it one more time since prepareAndSetAssetMetric is called in setAssetsMetrics > setMetrics?

File: app/process/watcher/prometheus/watcher.go (hunk 23 above)
Comment (nnedkov-limechain) [reply app/process/watcher/prometheus/watcher.go:71]: Ouh, yep will remove it.

File: cmd/clients.go [hunk 33]
```diff
@@ -48,9 +56,15 @@ func PrepareClients(config config.Clients) *Clients {
 		EVMClients[configChainId].SetChainID(clientChainId.Uint64())
 	}
 
+	logger := config.GetLoggerFor("Clients")
+
```
Comment (failfmi) [inline cmd/clients.go:60]: We do not need logging here. We use the Clients struct only to initialise stuff.

File: app/services/contracts/service.go [hunk 34]
```diff
@@ -203,3 +199,12 @@ func (bsc *Service) RemoveDecimals(amount *big.Int, asset string) (*big.Int, err
 	}
 	return amount, nil
 }
+
+func (bsc *Service) TokenFeeData() (struct {
```
Comment (failfmi) [inline app/services/contracts/service.go:203]: `TokenFeeData` is not used.

File: app/process/watcher/price/watcher.go [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func (pw *Watcher) beginWatching() {
+	err := pw.pricingService.FetchAndUpdateUsdPrices(false)
+	if err != nil {
+		pw.logger.Errorf(err.Error())
+	}
+
+	time.Sleep(10 * time.Minute)
+}
```
Comment (failfmi) [inline app/process/watcher/price/watcher.go:51]: This is called only once, it should be done every ~ 10mins. Can we add logs every time it fetches the prices?

Comment (failfmi) [reply app/process/watcher/price/watcher.go:51]: Not sure where yet, but can we log all the prices when they're fetched as well?

File: app/model/pricing/pricing.go [hunk 36]
```diff
@@ -0,0 +1,11 @@
+package pricing
```
Comment (failfmi) [inline app/model/pricing/pricing.go:1]: missing license

File: app/helper/http/get.go [hunk 37]
```diff
@@ -0,0 +1,30 @@
+package http
```
Comment (failfmi) [inline app/helper/http/get.go:1]: missing license

File: go.mod [hunk 38]
```diff
@@ -13,10 +13,12 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.0.0
 	github.com/rs/cors v1.7.0
+	github.com/shopspring/decimal v0.0.0-20200227202807-02e2044944cc
 	github.com/sirupsen/logrus v1.4.2
 	github.com/stretchr/testify v1.7.0
 	golang.org/x/net v0.0.0-20210907225631-ff17edfbf26d
 	google.golang.org/protobuf v1.26.1-0.20210525005349-febffdd88e85
+	gopkg.in/inf.v0 v0.9.1
```
Comment (failfmi) [inline go.mod:21]: This is not used; `go mod tidy`

**Review by vectorman1 — APPROVED** [review]

File: app/clients/evm/contracts/router/diamond-router.go [hunk 39]
```diff
@@ -963,30 +963,6 @@ func (_Router *RouterCaller) TokenFeeData(opts *bind.CallOpts, _token common.Add
...
-// Solidity: function tokenFeeData(address _token) view returns(uint256 serviceFeePercentage, uint256 feesAccrued, uint256 previousAccrued, uint256 accumulator)
-func (_Router *RouterCallerSession) TokenFeeData(_token common.Address) (struct {
-	ServiceFeePercentage *big.Int
-	FeesAccrued          *big.Int
-	PreviousAccrued      *big.Int
-	Accumulator          *big.Int
-}, error) {
-	return _Router.Contract.TokenFeeData(&_Router.CallOpts, _token)
-}
-
```
Comment (failfmi) [inline app/clients/evm/contracts/router/diamond-router.go:989]: This is autogenerated - should not be deleted from here.

**Review by svetlio8 — APPROVED** [review]

File: app/process/watcher/evm/watcher.go [hunk 40]
```diff
@@ -486,8 +494,16 @@ func (ew *Watcher) handleLockLog(eventLog *router.RouterLock, q qi.Queue) {
 			return
 		}
 	}
-	if properAmount.Cmp(big.NewInt(0)) == 0 {
-		ew.logger.Errorf("[%s] - Insufficient amount provided: Event Amount [%s] and Proper Amount [%s].", eventLog.Raw.TxHash, eventLog.Amount, properAmount)
```
Comment (failfmi) [inline app/process/watcher/evm/watcher.go:490]: This has to be reverted back.

**Review by failfmi — APPROVED** [review]
